embeddings/hyperparameter_search/configspace.py

```diff
@@ -0,0 +1,191 @@
+import abc
+from abc import ABC
+from dataclasses import dataclass, field
+from typing import Any, Dict, Final, Tuple, Type, TypeVar, Union
+
+import optuna
+
+from embeddings.hyperparameter_search.parameters import ConstantParameter, SearchableParameter
+from embeddings.utils.utils import PrimitiveTypes
+
+Parameter = Union[SearchableParameter, ConstantParameter]
+ParsedParameters = TypeVar("ParsedParameters")
+SampledParameters = Dict[str, Union[PrimitiveTypes, Dict[str, PrimitiveTypes]]]
+
+
+class ConfigSpace(ABC):
+    def _parse_parameter(
+        self, param_name: str, trial: optuna.trial.Trial
+    ) -> Tuple[str, PrimitiveTypes]:
+        param: Parameter = self.__getattribute__(param_name)
+        if isinstance(param, SearchableParameter):
+            return param.name, trial._suggest(name=param.name, distribution=param.distribution)
+        elif isinstance(param, ConstantParameter):
+            return param.name, param.value
+        else:
+            raise ValueError(
+                f"Parameter type {type(param)} is not suported! "
+                "Supported types are: SearchableParameter and ConstantParameter"
+            )
+
+    def _map_parameters(
+        self, trial: optuna.trial.Trial, parameters_types: Dict[str, Any]
```

Comment:
can parameter names be passed to this method instead of parameters types?

Reply:
Yes only names should be passed. I will handle it in next refactor